AIOOBE in QualifiedTypeReference.getTypeBinding() for call hierarchy

trancexpress opened the pull request:
This change adjusts the QualifiedTypeReference created for a record to have matching 'sources' and 'poss' array lengths, preventing an AIOOBE in QualifiedTypeReference.getTypeBinding() due to mismatched lengths.

Fixes: #790

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. TODO: how to report security issues
-->

## What it does
<!-- Include relevant issues and describe how they are addressed. -->

## How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

iloveeclipse commented:
Looks like jenkins detected new deadlock in the indexer (not because of this PR). 
Most likely introduced by https://github.com/eclipse-jdt/eclipse.jdt.core/pull/69. 
We should investigate that hang in a dedicated issue.

iloveeclipse commented:
> Looks like jenkins detected new deadlock in the indexer (not because of this PR).

Simeon, I see that you've pushed new commit and moved tests around and they are green now. So the deadlock was reproducible and root cause in the test code itself?

trancexpress commented:
> Simeon, I see that you've pushed new commit and moved tests around and they are green now. So the deadlock was reproducible and root cause in the test code itself?

I dont think so, I moved the test because its better placed in the Java 16 tests. I assumed the deadlock was sporadic.

iloveeclipse commented:
I've created https://github.com/eclipse-jdt/eclipse.jdt.core/issues/797 for the sporadic hang.

gayanper commented on org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/JavaSearchBugs16Tests.java:
```diff
@@ -621,6 +623,41 @@ public void testBug574870_4() throws CoreException {
 		}
+
+	/*
+	 * An inner record in a JRE class causes an AIOOBE during a call hierarchy computation.
```
Could we improve the comment here ? Even though description mentions about call hierarchy, we are testing a reference search isn't it ? And also the name `RecordClass` is not clear here since there is no java record involvement.

trancexpress commented:
Are we merging this @iloveeclipse ? Code freeze seems to be over (or is this again some problem with the code freeze check?).

iloveeclipse commented:
> Code freeze seems to be over 

Have you seen **any** mail saying that master is open for 4.28?

> (or is this again some problem with the code freeze check?).

Yes.

iloveeclipse commented:
**Now** we can merge :-)
